guides/10_other-tutorials/building-an-mcp-client-with-gradio.md

```diff
@@ -0,0 +1,489 @@
+# Using the Gradio Chatbot as an MCP Client
+
+This guide will walk you through a Model Context Protocol (MCP) Client and Server implementation with Gradio. You'll build a Gradio chatbot that can respond as an LLM (using Anthropic's Claude API) as well as generate images (thanks to a separate Gradio MCP Server).
+
+<video src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/gradio-guides/mcp-guides.mp4" style="width:100%" controls preload> </video>
+
+## What is MCP?
+
+The Model Context Protocol (MCP) standardizes how applications provide context to LLMs. It allows Claude to interact with external tools, like image generators, file systems, or APIs, etc.
+
+## Prerequisites
+
+- Python 3.10+
+- An Anthropic API key
+- Basic understanding of Python programming
+
+## Setup
+
+First, install the required packages:
+
+```bash
+pip install gradio anthropic mcp
+```
+
+Create a `.env` file in your project directory and add your Anthropic API key:
+
+```
+ANTHROPIC_API_KEY=your_api_key_here
+```
+
+## Part 1: Building the MCP Server
+
+The server provides tools that Claude can use. In this example, we'll create a server that generates images through [a HuggingFace space](https://huggingface.co/spaces/ysharma/SanaSprint).
+
+Create a file named `gradio_mcp_server.py`:
+
+```python
+from mcp.server.fastmcp import FastMCP
+import requests
+import json
+import sys
+import io
+import time
+
+# Set UTF-8 encoding for stdout/stderr
+sys.stdout = io.TextIOWrapper(sys.stdout.buffer, encoding='utf-8', errors='replace')
+sys.stderr = io.TextIOWrapper(sys.stderr.buffer, encoding='utf-8', errors='replace')
+
+# Initialize FastMCP server
+mcp = FastMCP("huggingface_spaces_image_display")
+
+@mcp.tool()
+async def generate_image(prompt: str, width: int = 512, height: int = 512) -> str:
```

Comment:
Does this have to be async? requests is synchronous so best to make this function synchronous as well or use a non-blocking http client like httpx

Reply:
I have not run this so in practice it may not block for a local case with one user. But still seems like best practice to make this function synchronous.